### PR TITLE
Implement df::coord iterator and cuboid class 

### LINKED ISSUE #4595: Cuboid struct/class for df::coord bounds
Testing if a `df::coord` is inside a given box is an issue that comes up when dealing with map positions. There doesn't seem to be anything handy in the DFHack library, so every tool has to implement its own logic.

Not sure where the struct should be defined. `MiscUtils`?

I created an example while working on an improved `plugins/plants.cpp`:
```
struct cuboid
{
    int16_t x_min = -1;
    int16_t x_max = -1;
    int16_t y_min = -1;
    int16_t y_max = -1;
    int16_t z_min = -1;
    int16_t z_max = -1;

    bool isValid() const
    {   // True if all max >= min >= 0
        return (x_min >= 0 && y_min >= 0 && z_min >= 0 &&
            x_max >= x_min && y_max >= y_min && z_max >= z_min);
    }

    bool addPos(int16_t x, int16_t y, int16_t z)
    {   // Expand cuboid to include point. Return true if bounds changed
        if (x < 0 || y < 0 || z < 0 || (isValid() && containsPos(x, y, z)))
            return false;

        x_min = (x_min < 0 || x < x_min) ? x : x_min;
        x_max = (x_max < 0 || x > x_max) ? x : x_max;

        y_min = (y_min < 0 || y < y_min) ? y : y_min;
        y_max = (y_max < 0 || y > y_max) ? y : y_max;

        z_min = (z_min < 0 || z < z_min) ? z : z_min;
        z_max = (z_max < 0 || z > z_max) ? z : z_max;

        return true;
    }
    inline bool addPos(const df::coord &pos) { return addPos(pos.x, pos.y, pos.z); }

    bool containsPos(int16_t x, int16_t y, int16_t z) const
    {   // Return true if point inside cuboid. Make sure cuboid is valid first!
        return x >= x_min && x <= x_max &&
            y >= y_min && y <= y_max &&
            z >= z_min && z <= z_max;
    }
    inline bool containsPos(const df::coord &pos) const { return containsPos(pos.x, pos.y, pos.z); }
};
```
This could probably be refined. Maybe add a constructor, a `clear()` function, and get/set (so we can keep track of isValid state.) I don't know if there's a use-case for supporting negative values (e.g., relative coordinates,) but this implementation doesn't.

Some existing plugins that could make use of this include `plugins/blueprint.cpp`, `plugins/dig-now.cpp`, and `plugins/regrass.cpp`. `Units::isUnitInBox` could be made to support this. `plugins/cleaners.cpp` could have this as an alternative to cleaning the entire map. `gui/teleport.lua` could probably make use of this if Lua doesn't make that a pain.

### LINKED ISSUE #4596: coordRange df::coord iterator for Maps module
Define a `df::coord` iterator `coordRange` for the `Maps` module, similar to `Units` module's `citizensRange`.

This should iterate all `df::coord`s contained in a cuboid defined by two given `df::coord`s. We might decide to not check if the input coords lie in the current loaded map bounds, because the function could be used to iterate block, region, etc., coords. (We can have a version that accepts a `df::coord2d` by setting `z` to 0 and returning the appropriate type.)

We can also accept a cuboid struct (#4595) and make use of it to simplify the iterator (rather than figure out which direction we need to iterate x/y/z.)

### PR DESCRIPTION
**Pull request by Bumber64**
Implement an iterator that operates on all `df::coord`s within a cuboid box. Iterates in the direction of x/y/z provided (rather than always min to max.) By default iterates per z-level, but has an option to iterate in the z direction first.
Closes #4596

Cuboid class for defining a cuboid of tiles. Can test if a coord is inside itself. Can be expanded to contain a given coord. Has a method to call the above iterator using the defined cuboid.
Closes #4595

Also update plugins to use `plant_type` enum attrs, since we're removing the `cuboid` struct from `plugins/plant.cpp` in this PR anyway.

### CODE REVIEW

**Comment by Bumber64** [comment]
Need to apply this to existing functions (e.g., `Units::isUnitInBox` and `Units::getUnitsInBox`).

File: library/include/modules/Maps.h [hunk 1]
```diff
@@ -180,6 +180,49 @@ namespace DFHack
...
+
+        // Expand cuboid to include point. Return true if bounds changed
+        DFHACK_EXPORT bool addPos(int16_t x, int16_t y, int16_t z);
+        DFHACK_EXPORT bool addPos(const df::coord &pos) { return addPos(pos.x, pos.y, pos.z); }
+
+        // Return true if point inside cuboid. Make sure cuboid is valid first!
+        DFHACK_EXPORT bool containsPos(int16_t x, int16_t y, int16_t z) const;
+        DFHACK_EXPORT bool containsPos(const df::coord &pos) const { return containsPos(pos.x, pos.y, pos.z); }
+
+        /// Iterate over every point in the cuboid from N-S, W-E, then top-down. Doesn't guarantee valid map tile!
```
Comment (myk002) [inline library/include/modules/Maps.h:222]: a `clamp()` function would be useful here to ensure that the cuboid is completely within valid map range.

File: library/include/modules/Maps.h [hunk 2]
```diff
@@ -180,6 +180,49 @@ namespace DFHack
...
+        /// If z_first iterate top-down, N-S, then W-E.
```
Comment (myk002) [inline library/include/modules/Maps.h:223]: so far, I've seen code that:
- doesn't care about whether z goes up or down
- requires z to go down (highest z-level to lowest)
so the options you have are likely ok. I don't know if there are any use cases for requiring that z goes from lowest to highest.

File: library/include/modules/Maps.h [hunk 3]
```diff
@@ -180,6 +180,49 @@ namespace DFHack
...
+        DFHACK_EXPORT void forCoord(std::function<void(df::coord)> fn, bool z_first = false);
```
Comment (myk002) [inline library/include/modules/Maps.h:224]: how hard would it be to provide a range interface? I added one for `Units::citizensRange`, and I like how it allows you to break out of the for loop early if you need to. `forCoord` would require that you iterate through all points (or you could change the signature of the callback to return a boolean that would indicate whether to exit early). I guess either option would be fine.

File: library/modules/Maps.cpp [hunk 4]
```diff
@@ -105,11 +105,92 @@ const char * DFHack::sa_feature(df::feature_type index)
...
+    if (z_first)
+    {   // Process z-columns first
+        for (int16_t x = x1; x != x2 + dx; x += dx)
+            for (int16_t y = y1; y != y2 + dy; y += dy)
+                for (int16_t z = z1; z != z2 + dz; z += dz)
+                    fn(df::coord(x, y, z));
+    }
+    else
+    {   // One z-level at a time
+        for (int16_t z = z1; z != z2 + dz; z += dz)
```
Comment (myk002) [inline library/modules/Maps.cpp:187]: most (all?) existing code does z first. does it make any sort of performance difference? is there any measurable advantage to be acting on the same map block in batches?

File: library/include/modules/Maps.h (hunk 3 above)
Comment (Bumber64) [reply library/include/modules/Maps.h:224]: Not quite sure how that works. `Units::citizensRange` filters a vector, but we're generating values from nowhere so I need some kind of example.

**Comment by Bumber64** [comment]
Should a class with only public members more properly be a struct, or is it weirder to have a struct with member functions?

File: library/include/modules/Maps.h (hunk 3 above)
Comment (myk002) [reply library/include/modules/Maps.h:224]: I couldn't find a good example, and buffering the whole sequence in a vector just so you can iterate through them seems like a waste. The second option might be better -- return a boolean from the callback to determine whether you should break out of the for loop early.

**Comment by myk002** [comment]
> Should a class with only public members more properly be a struct, or is it weirder to have a struct with member functions?

I don't have an opinion one way or the other here. I'd say do whatever feels natural.

File: library/include/modules/Maps.h (hunk 3 above)
Comment (Bumber64) [reply library/include/modules/Maps.h]: Tried `break`ing out of the nested for loop. Took me a while to realize the problem with that.

**Review by myk002 — APPROVED** [review]

**Pull request merged**